File: matirx_multiplication/bucket2/Matrix_double.py
import time
import numpy as np


def matrix_multiply_double(N, matrix1, matrix2):
    result = np.zeros((N, N), dtype=float)
    for i in range(N):
        for j in range(N):
            for k in range(N):
                result[i][j] += matrix1[i][k] * matrix2[k][j]
    return result

cpu_start_time = time.time()

N = 512
matrix1d = np.ones((N, N), dtype=float)
matrix2d = np.ones((N, N), dtype=float) * 2.0

meat_start_time = time.time()
matrix_multiply_double(N, matrix1d, matrix2d)
meat_time = time.time() - meat_start_time
print(f"Matrix size: {N}x{N} | Meat time (Double): {meat_time:.6f} seconds")

cpu_time = time.time() - cpu_start_time
print(f"Matrix size: {N}x{N} | CPU Time (Double): {cpu_time:.6f} seconds")

File: matirx_multiplication/Bucket1/Matrix_double.cpp
#include <iostream>
#include <ctime>
#include <vector>

using namespace std;


void matrixMultiplyDouble(int N, const vector<vector<double>> &matrix1, const vector<vector<double>> &matrix2, vector<vector<double>> &result) {
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            result[i][j] = 0.0;
            for (int k = 0; k < N; ++k) {
                result[i][j] += matrix1[i][k] * matrix2[k][j];
            }
        }
    }
}

double calculateElapsedTime(struct timespec &start, struct timespec &end) {
    long seconds = end.tv_sec - start.tv_sec;
    long nanoseconds = end.tv_nsec - start.tv_nsec;
    return seconds + nanoseconds * 1e-9;
}

int main() {
        
    struct timespec cpustart,cpuend;
    clock_gettime(CLOCK_MONOTONIC, &cpustart);

    int N = 1024;
    vector<vector<double>> matrix1(N, vector<double>(N, 1));
    vector<vector<double>> matrix2(N, vector<double>(N, 2));
    vector<vector<double>> result(N, vector<double>(N, 0));

    struct timespec meatstart, meatend;
    clock_gettime(CLOCK_MONOTONIC, &meatstart);

    matrixMultiplyDouble(N, matrix1, matrix2, result);

    clock_gettime(CLOCK_MONOTONIC, &meatend);

    double MeatTime = calculateElapsedTime(meatstart, meatend);
    cout << "Matrix size: " << N << "x" << N << " | Meat time (Double): " << MeatTime << " seconds" << endl;

    clock_gettime(CLOCK_MONOTONIC, &cpuend);

    double cpuTime = calculateElapsedTime(cpustart, cpuend);
    cout << "Matrix size: " << N << "x" << N << " | Elapsed time (Double): " << cpuTime << " seconds" << endl;

    return 0;
}
